core/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/ThrottlingAllocationDecider.java

```diff
@@ -102,17 +108,16 @@ public Decision canAllocate(ShardRouting shardRouting, RoutingNode node, Routing
 
     @Override
     public Decision canAllocate(RoutingNode node, RoutingAllocation allocation) {
-        int currentRecoveries = 0;
-        for (ShardRouting shard : node) {
-            if (shard.initializing()) {
-                currentRecoveries++;
-            }
-        }
-        if (currentRecoveries >= concurrentRecoveries) {
+        int currentOutRecoveries = allocation.routingNodes().getOutgoingRecoveries(node.nodeId());
+        int currentInRecoveries = allocation.routingNodes().getIncomingRecoveries(node.nodeId());
+        if (currentOutRecoveries >= concurrentOutgoingRecoveries) {
+            return allocation.decision(Decision.THROTTLE, NAME, "too many shards currently recovering [%d], limit: [%d]",
+                currentOutRecoveries, concurrentOutgoingRecoveries);
+        } else if (currentInRecoveries >= concurrentIncomingRecoveries) {
             return allocation.decision(Decision.THROTTLE, NAME, "too many shards currently recovering [%d], limit: [%d]",
-                    currentRecoveries, concurrentRecoveries);
-        } else {
-            return allocation.decision(Decision.YES, NAME, "below shard recovery limit of [%d]", concurrentRecoveries);
+                currentInRecoveries, concurrentIncomingRecoveries);
+        }  else {
+            return allocation.decision(Decision.YES, NAME, "below shard recovery limit of [%d]", concurrentOutgoingRecoveries);
```

Comment:
Should probably mention both limits here, or else it could be confusing if they are different values